### ISSUE TITLE
Writing a controlled component

### ISSUE THREAD

**Issue by guyzmo** [keajs/kea#79]
Hello and thanks for the work on kea,

I've been using Kea to write a webapp, and so far it's been quite a pleasure. But as I've been learning kea while writing the code, I'm trying to refactor it to fix the ugly/bunky bits of my code into something more reusable. And something that I don't understand from kea's documentation is how I can build controlled components, such as:

```
<ThatComponent id={'fubar'} prop1={42} prop2={'don't panic'} />
```

how can I reason on `prop1`/`prop2` (as being part of another store) within `ThatComponent`'s `kea()` parts? I could do a `mapPropsToState` and just forget about `kea`, but until now, I've been delighted with writing stores with kea, and I'd like to factorize my code and keep using it.

**Comment by mariusandra** [comment]
Hey, can you give a small example of what prop1 and prop2 will be doing?

**Comment by guyzmo** [comment]
hm… let's consider a  better usage example, I'm updating the question with it:

```jsx
@kea({
  /* let's consider we have data1 and data2 defined as reducers here */
})
class SomeForm extends Component {
  render () {
    const {
      data1,
      data2
    } = this.props;

    return (<div>
              <!-- some stuff here -->
              <ThatComponent prop1={data1} prop2={data2} />
              <!-- some other stuff there -->
            </div>);

}
```

how do I create `ThatComponent` as a controlled component, so that I can update `data1` and `data2` in there, synced with the values in `SomeForm`?

```
@kea({
  /* I want prop1 and prop2 as reducers, or as base for selectors */
})
class ThatComponent extend Component {
  render () {
    return (
      <!-- stuff that shows off prop1 and prop2 directly or as a derivation of them and/or calls actions on them to modify them -->
    )
  }
}
```

I'm trying to keep the example generic to pinpoint what I don't understand and avoid bloating it with specifics of my use case, I hope it's clear ☺

N.B.: if you're hanging on some chat system, I'd be very happy to join to better help and get help on those things ☺ (I still like IRC and the art of chatting).

**Comment by guyzmo** [comment]
So to try to achieve this on my own, here's what I did:

As a first step, I wanted to just show transformations applied to the props, and looking through the ticketing I found that I could refer to some given props within reducers the following way: 

```jsx
@kea({
  key: (props) => props.id,
  path: (key) => ['foo', 'bar', key],
  selectors: ({selectors}) => ({
    transformed_1: [
      () => [(_, props) => props.prop1],
      (that_prop_1) => `This is a transformation of #{prop1}!`,
      PropTypes.string
    ],
    transformed_2: [
      () => [(_, props) => props.prop1],
      (that_prop_1) => `This is a transformation of #{prop1}!`,
      PropTypes.string
    ]
})
class ThatComponent extends Component {
  render () {
    return (
      <div>
        <div>{this.props.transformed_1}</div>
        <div>{this.props.transformed_1}</div>
      </div>
    );
  }
}
```

but it looks like this is not the way to go, because I'm getting an infinite rendering loop... And anyway, it's only useful for transformations of props given by the outter component to the inner one. So the idea is barely useful, and doesn't anyway work.

What I'd really like would be to have the prop to be a local state, so that I can define actions and reducers with kea for the inner component `ThatComponent` and it updates the outter component's state.

Maybe am I wanting something I shouldn't do with kea, and I shall instead just do the usual state/props mapping? But if it can be done, I really don't get how to achieve that.

**Comment by xantorres** [comment]
For what you say, it looks like you either need to pass the container actions as props to the children, or to use `connect` in the children.

Just create a `kea` controller (without `key`) in the container which exposes `props` or `actions` and then use `connect` in the inner children components, or if you prefer, just pass the parent actions and re-selected values to the children via props.

Either way, if you pretend to keep your children components dumb (or controlled), they shouldn't have any logic or even be connected to the store, (unless you want to get some logic from somewhere else using connect). Most of the stuff should just come from the props.

PS: Btw, the 2 selectors you show in your example are using the same input prop, could that be the reason you are getting an infinite loop?

**Comment by guyzmo** [comment]
So (to give some more context beyond the generic example) my concrete use case is that I'm defining a datetime component that exposes a list of dates and time given some business logic (only a time between 7am and 8pm). So if current date is today, and current time is between 7am and 8pm, it exposes only time started next hour up to 8pm, if time is or is after 8pm, the date is the next day.

I wrote the logic in the `render()` function, and it works, but I found that solution rather inelegant. The stored data is aimed to be used in a larger form, and as I'm building a form, I'd like to have the date and time managed by the component but its result exposed to the outter context with minimal code and greater readability. I'd also like to have that component used in other places, removing the tight coupling.

> For what you say, it looks like you either need to pass the container actions as props to the children, or to use connect in the children.

I'm currently passing the date (`prop1`) and time (`prop2`) states and actions as props to the component (`ThatComponent`) and do what needs to be done from within the container. But something feels wrong about that solution, making the component too tightly coupled to the context.

I'm not sure to understand your thoughts using connect? How would that be like?

> PS: Btw, the 2 selectors you show in your example are using the same input prop, could that be the reason you are getting an infinite loop?

that was a mistake when I wrote the example, I corrected that, thanks :)

**Comment by mariusandra** [comment]
Hey, you sort of have two standard paths here:
1) a component that's uncoupled, receives `date`, `time` and `onChange` as props and does what it needs to with all of them
2) a component that's tightly coupled, is itself `@connect`ed and knows where to get its props and what actions to call.

I'm not sure what is the third way that you're trying to reach :).

If I understand correctly, the issue is that you want the inner component to perform 2 functions:
1) save the updated state from when an action is performed (someone changes the date or time)
2) sanitise the input from the outer component (today 10pm = tomorrow 7am) and save it as needed

In this case you have 3 options:
1) If you don't want to save the sanitised state (you'll get bad input until manually changed), just clean it up in the inner component's `render` as you do now. Perhaps extract that code into a pure helper function that you could import and run from other parts of your app when needed.
2) In the inner component's `componentDidMount` and `componentWillReceiveProps`, check and sanitise the times and call the given `onChange` function to save the sanitised state.
3) Make sure the data passed to the component is always sanitised, so move the helper function one level higher and put it somewhere in your data model, e.g. the code that creates the data for the initial form... and/or the kea/redux reducer that acts on the action that gives it the new date.

Not sure if this was of much help. If not, please give a more concrete example. It's hard to follow with the `prop1`s and `prop2`s, sorry :).
